### PR TITLE
Adds support for deferrable foreign key constraints in PostgreSQL

### PR DESCRIPTION
**Pull request by benedikt**
### Summary

By default, foreign key constraints in PostgreSQL are checked after each statement. This works for most use cases, but becomes a major limitation when creating related records before the parent record is inserted into the database.

One example of this is looking up / creating a person via one or more unique alias.

```ruby
Person.transaction do
  alias = Alias
    .create_with(user_id: SecureRandom.uuid)
    .create_or_find_by(name: "DHH")

  person = Person
    .create_with(name: "David Heinemeier Hansson")
    .create_or_find_by(id: alias.user_id)
end
```

Using the default behavior, the transaction would fail when executing the first `INSERT` statement.

This pull request adds support for deferrable foreign key constraints by adding a new option to the `add_foreign_key` statement in migrations:

```ruby
add_foreign_key :aliases, :person, deferrable: true
```

The `deferrable: true` leaves the default behavior, but allows manually deferring the checks using `SET CONSTRAINTS ALL DEFERRED` within a transaction. This will cause the foreign keys to be checked after the transaction.

It's also possible to adjust the default behavior from an immediate check (after the statement), to a deferred check (after the transaction).

```ruby
add_foreign_key :aliases, :person, deferrable: :deferred
```

### Other Information

There is a pull request #17094 from 2014 that implements generic SQL options on `add_foreign_key`. Pull request #40192 recently implemented the validate aspect of that.

### CODE REVIEW

**Comment by zzak** [comment]
@benedikt I will have to do some more research on this but is this something you're still interested in picking up?

**Comment by benedikt** [comment]
@zzak Yes, absolutely! Let me know how I can help with research.

**Comment by zzak** [comment]
@benedikt First thing I notice is that C[I failed](https://buildkite.com/rails/rails/builds/77780#73756c8b-8385-4d03-a6c7-bfc542cea896), but I'm not entirely sure it is related yet.

```
Failure:
--
&nbsp; | ActiveRecord::ConnectionAdapters::ConnectionPoolTest#test_checkout_fairness [/rails/activerecord/test/cases/connection_pool_test.rb:395]:
&nbsp; | --- expected
&nbsp; | +++ actual
&nbsp; | @@ -1 +1 @@
&nbsp; | -[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
&nbsp; | +[1, 2, 3, 4, 5, 6, 10, 7, 8, 9]
```

Will have to take a closer look at this, thank you for your patience :bow:

**Comment by benedikt** [comment]
@zzak Just merged the current main into this branch to resolve conflicts which caused the CI to fail. I don't think this failure is related, but I can look into it more.

**Comment by benedikt** [comment]
@zzak Tried to reproduce this locally but didn't see the failure there. 

After rebasing, the CI also [didn't complain](https://buildkite.com/rails/rails/builds/77785#0c90959a-be60-4570-bf4c-7d62d38758e2) didn't complain about it again. 

As the [failure](https://buildkite.com/rails/rails/builds/77780#73756c8b-8385-4d03-a6c7-bfc542cea896) was in the Sqlite3 tests and this pull request only touches PostgreSQL, I think it's save to assume it was an unrelated failure. 

There now seems to be a failure related to memcache timeout, but I guess that's also not related to the changes of this pull request.

**Comment by benedikt** [comment]
@zzak It's a miracle, but all tests are passing. Even the unrelated ones. 😄

File: activerecord/CHANGELOG.md
Comment (ghiculescu) [inline activerecord/CHANGELOG.md:3]: Probably don't need this line.

It would be cool to see an example here though.

File: activerecord/test/cases/migration/foreign_key_test.rb
Comment (ghiculescu) [inline activerecord/test/cases/migration/foreign_key_test.rb:522]: can you please add schema dump tests for `deferrable: :immediate` and `deferrable: false`

Comment (benedikt) [reply activerecord/test/cases/migration/foreign_key_test.rb:540]: Just added tests for both cases, as well as a test to make sure the `deferrable: false` doesn't do anything in the database itself.

File: activerecord/CHANGELOG.md
Comment (benedikt) [reply activerecord/CHANGELOG.md]: Yeah, you're right. Added the example from the pull request description.

**Comment by benedikt** [comment]
Thanks for your review, @ghiculescu. As mentioned in the individual comments, I adjusted the pull request accordingly. 

Unfortunately, the build once again isn't green because of [an HTTP timeout](https://buildkite.com/rails/rails/builds/78047#f7b9e336-ddbd-4b28-9a34-592647cc7b6b) in an unrelated part.

**Review by ghiculescu — APPROVED** [review]
This looks great to me, I will wait for someone from the core team to review. Thanks for your work on this!

**Review by zzak — APPROVED** [review]

**Comment by zzak** [comment]
@benedikt You will have to squash the commits but maybe better to wait until another committer reviews this 🙏

**Comment by benedikt** [comment]
@zzak Yup, it's already just one commit.

**Comment by benedikt** [comment]
@zzak @ghiculescu Anything I can do to move this forward? 😅

**Comment by ghiculescu** [comment]
There’s a merge conflict, other than that it’s waiting on the core team to review.

**Comment by benedikt** [comment]
@ghiculescu Yeah, I keep fixing that particular merge conflict for the last couple of months. How do we get the core team to review this? 😄

**Comment by ghiculescu** [comment]
Patience is the name of the game :) There's lots of open PRs and they're all pretty busy with Rails + real jobs. They do eventually get to most PRs but you just have to wait it out. The best advice I've heard is "make the PR easy to merge", which I think you have done a good job of here.

**Comment by benedikt** [comment]
@rafaelfranca Thank you for merging this 👍

**Pull request merged**
